modules/e2ee/OlmAdapter.js

```diff
@@ -470,6 +578,201 @@ export class OlmAdapter extends Listenable {
+                    sas,
+                    transactionId,
+                    isInitiator: false
+                };
+
+                const pubKey = olmData.sasVerification.sas.get_pubkey();
+                const commitment = this._computeCommitment(pubKey, msg.data);
+
+                /* The first phase of the verification process, the Key agreement phase
+                   https://spec.matrix.org/latest/client-server-api/#short-authentication-string-sas-verification
+                */
+                const acceptMessage = {
+                    [JITSI_MEET_MUC_TYPE]: OLM_MESSAGE_TYPE,
+                    olm: {
+                        type: OLM_MESSAGE_TYPES.SAS_ACCEPT,
+                        data: {
+                            transactionId,
+                            commitment
+                        }
+                    }
+                };
+
+                this._sendMessage(acceptMessage, pId);
+            } else {
+                logger.debug(`Received sas init message from ${pId} but we have no session for them!`);
+
+                this._sendError(participant, 'No session found while processing sas-init');
+            }
+            break;
+        }
+        case OLM_MESSAGE_TYPES.SAS_ACCEPT: {
+            if (olmData.session) {
+                const { commitment, transactionId } = msg.data;
+
+                olmData.sasVerification.sasCommitment = commitment;
+
+                const pubKey = olmData.sasVerification.sas.get_pubkey();
+
+                // Send KEY.
+                const keyMessage = {
+                    [JITSI_MEET_MUC_TYPE]: OLM_MESSAGE_TYPE,
+                    olm: {
+                        type: OLM_MESSAGE_TYPES.SAS_KEY,
+                        data: {
+                            key: pubKey,
+                            transactionId
+                        }
+                    }
+                };
+
+                this._sendMessage(keyMessage, pId);
+
+                olmData.sasVerification.keySent = true;
+            } else {
+                logger.debug(`Received sas accept message from ${pId} but we have no session for them!`);
+
+                this._sendError(participant, 'No session found while processing sas-accept');
+            }
+            break;
+        }
+        case OLM_MESSAGE_TYPES.SAS_KEY: {
+            if (olmData.session) {
+                const { isInitiator, sas, sasCommitment, startContent, keySent } = olmData.sasVerification;
+
+                if (sas.is_their_key_set()) {
+                    logger.warn('SAS already has their key!');
+
+                    return;
+                }
+
+                const { key: theirKey, transactionId } = msg.data;
+
+                if (sasCommitment) {
+                    const commitment = this._computeCommitment(theirKey, startContent);
+
+                    if (sasCommitment !== commitment) {
+                        this._sendError(participant, 'OlmAdapter commitments mismatched');
```

Comment:
rather than a generic error we should send the sas completed event with an error indicator